File: scripts/tools/twitter_tools.py
"""Twitter tools - wrappers for Node.js Twitter API endpoints using CrewAI tool format."""
from typing import Any, Dict, List, Optional
from crewai.tools import tool
from .http_client import call_nodejs_tool


@tool("Post a tweet")
def post_tweet(text: str, media_ids: Optional[List[str]] = None) -> str:
    """Post a tweet to Twitter. Returns JSON result string."""
    result = call_nodejs_tool("/api/tools/twitter/post", data={
        "text": text,
        "media_ids": media_ids,
    })
    return str(result)


@tool("Reply to a tweet")
def reply_to_tweet(text: str, tweet_id: str) -> str:
    """Reply to a tweet on Twitter. Returns JSON result string."""
    result = call_nodejs_tool("/api/tools/twitter/reply", data={
        "text": text,
        "tweet_id": tweet_id,
    })
    return str(result)


@tool("Get Twitter mentions")
def get_mentions(max_results: int = 10, since_id: Optional[str] = None) -> str:
    """Get mentions timeline from Twitter. Returns JSON result string."""
    params = {"max_results": max_results}
    if since_id:
        params["since_id"] = since_id
    result = call_nodejs_tool("/api/tools/twitter/mentions", method="GET", params=params)
    return str(result)


@tool("Like a tweet")
def like_tweet(tweet_id: str) -> str:
    """Like a tweet on Twitter. Returns JSON result string."""
    result = call_nodejs_tool("/api/tools/twitter/like", data={"tweet_id": tweet_id})
    return str(result)


@tool("Retweet a tweet")
def retweet_tweet(tweet_id: str) -> str:
    """Retweet a tweet on Twitter. Returns JSON result string."""
    result = call_nodejs_tool("/api/tools/twitter/retweet", data={"tweet_id": tweet_id})
    return str(result)


# Alias for backward compatibility
retweet = retweet_tweet


@tool("Search tweets")
def search_tweets(query: str, max_results: int = 10) -> str:
    """Search for tweets on Twitter. Returns JSON result string."""
    result = call_nodejs_tool("/api/tools/twitter/search", data={
        "query": query,
        "max_results": max_results,
    })
    return str(result)


@tool("Quote tweet")
def quote_tweet(text: str, tweet_id: str) -> str:
    """Quote a tweet on Twitter. Returns JSON result string."""
    result = call_nodejs_tool("/api/tools/twitter/quote", data={
        "text": text,
        "tweet_id": tweet_id,
    })
    return str(result)


@tool("Get tweet details")
def get_tweet_details(tweet_id: str) -> str:
    """Get details of a tweet. Returns JSON result string."""
    result = call_nodejs_tool(f"/api/tools/twitter/tweet/{tweet_id}", method="GET")
    return str(result)


@tool("Get trending topics")
def get_trending_topics(woeid: int = 1) -> str:
    """Get trending topics (placeholder - Twitter API v2 requires elevated access). Returns JSON result string."""
    # Twitter API v2 doesn't have trends endpoint for free tier
    return str([{"error": "Trends API requires elevated Twitter access"}])

File: src/tools/twitterTools.ts
/** Twitter API Tools — ported from twitter_tools.py using twitter-api-v2. */
import { TwitterApi } from "twitter-api-v2";
import { settings } from "../config/settings.js";

function getTwitterClient(): TwitterApi {
  return new TwitterApi({
    appKey: settings.twitter.apiKey,
    appSecret: settings.twitter.apiSecret,
    accessToken: settings.twitter.accessToken,
    accessSecret: settings.twitter.accessSecret,
  });
}

function getReadOnlyClient(): TwitterApi {
  return new TwitterApi(settings.twitter.bearerToken);
}

export async function postTweet(
  text: string,
  mediaIds?: string[]
): Promise<Record<string, unknown>> {
  const client = getTwitterClient();
  try {
    const params: Record<string, unknown> = { text };
    if (mediaIds?.length) {
      params.media = { media_ids: mediaIds };
    }
    const result = await client.v2.tweet(params as any);
    const tweetId = result.data.id;
    return { success: true, tweet_id: tweetId, url: `https://twitter.com/i/status/${tweetId}` };
  } catch (e: any) {
    return { success: false, error: e.message };
  }
}

export async function replyToTweet(
  text: string,
  tweetId: string
): Promise<Record<string, unknown>> {
  const client = getTwitterClient();
  try {
    const result = await client.v2.reply(text, tweetId);
    const replyId = result.data.id;
    return { success: true, reply_id: replyId, url: `https://twitter.com/i/status/${replyId}` };
  } catch (e: any) {
    return { success: false, error: e.message };
  }
}

export async function getMentions(
  maxResults = 10,
  sinceId?: string | null
): Promise<Record<string, unknown>[]> {
  const client = getTwitterClient();
  try {
    const me = await client.v2.me();
    const userId = me.data.id;

    const params: Record<string, unknown> = {
      max_results: maxResults,
      "tweet.fields": ["author_id", "created_at", "conversation_id", "in_reply_to_user_id"],
      expansions: ["author_id", "referenced_tweets.id"],
    };
    if (sinceId) params.since_id = sinceId;

    const mentions = await client.v2.userMentionTimeline(userId, params as any);

    const results: Record<string, unknown>[] = [];
    if (mentions.data?.data) {
      for (const tweet of mentions.data.data) {
        results.push({
          tweet_id: tweet.id,
          text: tweet.text,
          author_id: tweet.author_id,
          created_at: tweet.created_at,
          conversation_id: tweet.conversation_id,
        });
      }
    }
    return results;
  } catch (e: any) {
    return [{ error: e.message }];
  }
}

export async function likeTweet(tweetId: string): Promise<Record<string, unknown>> {
  const client = getTwitterClient();
  try {
    const me = await client.v2.me();
    await client.v2.like(me.data.id, tweetId);
    return { success: true, tweet_id: tweetId };
  } catch (e: any) {
    return { success: false, error: e.message };
  }
}

export async function retweetTweet(tweetId: string): Promise<Record<string, unknown>> {
  const client = getTwitterClient();
  try {
    const me = await client.v2.me();
    await client.v2.retweet(me.data.id, tweetId);
    return { success: true, tweet_id: tweetId };
  } catch (e: any) {
    return { success: false, error: e.message };
  }
}

export async function getTweetDetails(tweetId: string): Promise<Record<string, unknown>> {
  const client = getReadOnlyClient();
  try {
    const result = await client.v2.singleTweet(tweetId, {
      "tweet.fields": ["author_id", "created_at", "public_metrics", "conversation_id"],
      expansions: ["author_id"],
    });

    if (result.data) {
      const metrics = (result.data.public_metrics || {}) as Record<string, number>;
      return {
        tweet_id: result.data.id,
        text: result.data.text,
        author_id: result.data.author_id,
        created_at: result.data.created_at,
        likes: metrics.like_count ?? 0,
        retweets: metrics.retweet_count ?? 0,
        replies: metrics.reply_count ?? 0,
        quotes: metrics.quote_count ?? 0,
      };
    }
    return { error: "Tweet not found" };
  } catch (e: any) {
    return { error: e.message };
  }
}

export async function getTrendingTopics(
  _woeid = 1
): Promise<Record<string, unknown>[]> {
  // Twitter API v2 doesn't have a direct trends endpoint for free tier.
  // This is a placeholder that returns an empty list.
  // In production, use the v1.1 endpoint or a third-party trends service.
  return [{ error: "Trends API requires elevated access" }];
}

export async function searchTweets(
  query: string,
  maxResults = 10
): Promise<Record<string, unknown>[]> {
  const client = getReadOnlyClient();
  try {
    const result = await client.v2.search(query, {
      max_results: maxResults,
      "tweet.fields": ["author_id", "created_at", "public_metrics"],
    });

    const tweets: Record<string, unknown>[] = [];
    if (result.data?.data) {
      for (const tweet of result.data.data) {
        tweets.push({
          tweet_id: tweet.id,
          text: tweet.text,
          author_id: tweet.author_id,
          created_at: tweet.created_at,
          public_metrics: tweet.public_metrics,
        });
      }
    }
    return tweets;
  } catch (e: any) {
    return [{ error: e.message }];
  }
}

export async function quoteTweet(
  text: string,
  tweetId: string
): Promise<Record<string, unknown>> {
  const client = getTwitterClient();
  try {
    const result = await client.v2.tweet({
      text,
      quote_tweet_id: tweetId,
    } as any);
    const newId = result.data.id;
    return { success: true, tweet_id: newId, url: `https://twitter.com/i/status/${newId}` };
  } catch (e: any) {
    return { success: false, error: e.message };
  }
}
